QMS Api structure with models, controllers and routes.

gdprasadAU opened the pull request.

gdprasadAU committed:
Added Structure

gdprasadAU committed:
Added DB modules and Plans API

gdprasadAU committed:
Added User group and resources.

gdprasadAU committed:
Added User routes and Api

gdprasadAU committed:
Added Admin routes and Api

gdprasadAU committed:
master

gdprasadAU committed:
Rebase

gdprasadAU committed:
Added admin to acces user plan detailsso that admin will be able to view user details.

gdprasadAU committed:
added Swagger doc

gdprasadAU committed:
Added swageer doc for Quotas,resources, usages.

slr71 commented on go.mod:
```diff
@@ -0,0 +1,40 @@
+
+require (
+	github.com/cyverse-de/echo-middleware v0.1.0
+	github.com/labstack/echo v3.3.10+incompatible
```
This version of Echo relies upon a JWT library that is unmaintained and contains a security vulnerability. Switching to version 4 isn't too painful, though. [This PR for cyverse-de/requests](https://github.com/cyverse-de/requests/pull/11) makes the switch and provides a brief overview of the steps. If you could switch QMS as well, that would be very helpful.

slr71 commented on Makefile:
```diff
@@ -0,0 +1,15 @@
+clean:
+	rm -rf swagger.json requests vendor
+
+.PHONY: install-swagger clean all
```
This Makefile doesn't quite work as it's currently written. It sounds like you've already got some changes in the works, though, so I won't make any specific suggestions yet. :)

slr71 commented on server/server.go:
```diff
@@ -0,0 +1,41 @@
+	// Establish the database connection.
+	log.Info("establishing the database connection")
+
+	databaseURI := "postgres://postgres:password@localhost:54320/qmsdb?sslmode=disable" //cfg.GetString("notifications.db.uri")
```
It'll be necessary to make this configurable before we can deploy the QMS into Kubernetes so that we can customize which database host the QMS connects to. The DE services all use configuration files for this purpose. CACAO uses https://github.com/kelseyhightower/envconfig instead, which is pretty nice, and might also make it a bit easier to deploy the QMS into Kubernetes.

slr71 commented on server/server.go:
```diff
@@ -0,0 +1,41 @@
+		Router:  e,
+		DB:      db,
+		GORMDB:  gormdb,
+		Service: "qms",
+		Title:   "serviceInfo.Title",   //TODO: correct this
+		Version: "serviceInfo.Version", //TODO:correct this
```
Other services load these values from the generated Swagger docs so that we only have to define the service name, title and version in one place. You can find some example code for it in [swagger_service_info.go](https://github.com/cyverse-de/requests/blob/main/swagger_service_info.go) in `cyverse-de/requests`.

slr71 commented on cmd/main.go:
```diff
@@ -0,0 +1,9 @@
+package main
```
I'd actually kind of prefer it if this file were placed in the main directory of the repository. That way, the executable file will automatically be named `QMS` and the Makefile would work with just a couple of tweaks.

slr71 commented on server/server.go:
```diff
@@ -0,0 +1,41 @@
+	}
+
+	// Register the handlers.
+	RegisterHandlers(s)
```
I like the idea of having a separate function for this. 👍

slr71 commented on server/router.go:
```diff
@@ -0,0 +1,87 @@
+	return e
+}
+
+// type SwaggerServiceInfo struct {
+// 	Description string `json:"description"`
+// 	Title       string `json:"title"`
+// 	Version     string `json:"version"`
+// }
+
+// // PartialSwagger is a structure that can be used to read just the service information from a Swagger JSON document.
+// type PartialSwagger struct {
+// 	Info *SwaggerServiceInfo `json:"info"`
+// }
+
+// func getSwaggerServiceInfo() (*SwaggerServiceInfo, error) {
+// 	wrapMsg := "unable to load the Swagger JSON"
+
+// 	// Open the file containing the Swagger JSON.
+// 	file, err := os.Open("swagger.json")
+// 	if err != nil {
+// 		return nil, errors.Wrap(err, wrapMsg)
+// 	}
+
+// 	// Parse the bits we want out of the Swagger JSON.
+// 	decoder := json.NewDecoder(file)
+// 	partialSwagger := &PartialSwagger{}
+// 	err = decoder.Decode(partialSwagger)
+// 	if err != nil {
+// 		return nil, errors.Wrap(err, wrapMsg)
+// 	}
+// 	return partialSwagger.Info, nil
+// }
```
Ah! It looks like this code was grabbed from [swagger_service_info.go](https://github.com/cyverse-de/requests/blob/main/swagger_service_info.go) in `cyverse-de/requests`. The code probably wasn't working before because `swagger.json` wasn't present in the current working directory. Now that the file is being generated, it might be easier to get this code working.

slr71 replied:
This code can probably also be removed until we have some more time to get it working.

slr71 commented on server/router.go:
```diff
@@ -0,0 +1,87 @@
+	//Plans
+	plans := v1.Group("/plans")
+	plans.GET("/plans", s.GetAllPlans)
+	plans.GET("/plans/:plan_id", s.GetPLansForID)
```
```suggestion
	plans.GET("", s.GetAllPlans)
	plans.GET("/:plan_id", s.GetPLansForID)
```

As it was written, it produced URL paths such as `/v1/plans/plans` and `/v1/plans/plans/some-plan-id`.

slr71 commented on server/router.go:
```diff
@@ -0,0 +1,87 @@
+// 	}
+// 	return partialSwagger.Info, nil
+// }
+func RegisterHandlers(s controllers.Server) {
```
There aren't many endpoints to add data to the database. It would be helpful to have some endpoints for adding new entries to the database.

slr71 commented on internal/controllers/admin.go:
```diff
@@ -0,0 +1,215 @@
+	now := time.Now().Format("2006-01-02")
+
+	plandata := []PlanDetails{}
+	err := s.GORMDB.Raw(`select plans.name,usage.usage,quotas.quota,resource_types.unit from
```
Looks like there's a typo here:

```suggestion
	err := s.GORMDB.Raw(`select plans.name,usages.usage,quotas.quota,resource_types.unit from
```

slr71 commented on server/router.go:
```diff
@@ -0,0 +1,87 @@
+	// serviceInfo, err := getSwaggerServiceInfo()
+	// if err != nil {
+	// 	e.Logger.Fatal(err, serviceInfo)
+	// }
```
This is the code that sets up the ReDoc server, which serves up documentation for the service. It's a really handy thing to have. Feel free to let me know if you have any questions about setting it up.

slr71 commented on internal/controllers/plans.go:
```diff
@@ -0,0 +1,39 @@
+	return ctx.JSON(http.StatusOK, model.SuccessResponse(data, http.StatusOK))
+}
+
+// swagger:route GET /plans/{PlanID} plans listPlansByID
+// Returns a List all the plans
```
```suggestion
//
// Get Plan Information
//
// Returns information about a the plan with the given identifier.
```

The first line of the comment is used as the summary. The second line is used as the description. For example, these are some comments from `cyverse-de/requests`:

```go
// swagger:route GET /request-types request-types getRequestTypes
//
// List Request Types
//
// This endpoint lists all of the request types that have been registered, sorted by name.
//
// responses:
//    200: requestTypeListing
//    500: errorResponse
```

The generated Swagger docs are rendered like this:

![image](https://user-images.githubusercontent.com/415143/148468653-b1ef3c0b-f2ae-4fbc-9946-9b2201b0ff07.png)

slr71 replied:
Another suggestion that I missed for this yesterday was to mention the function name in the comment. The reason for this is that the Go linter will complain about the comment if it doesn't begin with the function name. I also like the idea of not placing a space between the comment and the function definition. For example, something like this seems to work well:

```go
// GetAllPlans implements the GET /plans endpoint.
//
// swagger:route GET /plans plans listPlans
//
// List Plans
//
// Returns a List all the plans
//
// responses:
//   200: plansResponse
//   404: RootResponse
func (s Server) GetAllPlans(ctx echo.Context) error {
	data := []model.Plans{}
	err := s.GORMDB.Debug().Find(&data).Error
	if err != nil {
		return ctx.JSON(http.StatusInternalServerError, model.ErrorResponse(err.Error(), http.StatusInternalServerError))
	}
	return ctx.JSON(http.StatusOK, model.SuccessResponse(data, http.StatusOK))
}
```

Similarly, summary comments that begin with the function name should be added to all functions that begin with a capital letter. Otherwise, the Go linter will complain.

slr71 commented on internal/controllers/plans.go:
```diff
@@ -0,0 +1,39 @@
+	"github.com/labstack/echo"
+)
+
+// swagger:route GET /plans plans listPlans
+// Returns a List all the plans
+// responses:
+//   200: plansResponse
+//   404: RootResponse
```
```suggestion
// swagger:route GET /plans plans listPlans
//
// List Plans
//
// Returns a List of all the plans.
//
// responses:
//   200: plansResponse
//   404: RootResponse
```

See the comment for `GET /plans/{planID}` for an explanation.

slr71 reviewed (commented):
I haven't finished reviewing all of the code yet, but I wanted to make the comments available to you. This looks like a good start so far! Some overall suggestions that I have so far are:

- Switch to `labstack/echo` version 4 (version 3 depends on a JWT library with a security vulnerability).
- Finish documenting the endpoints so that they all show up in the generated Swagger docs.
- Update the existing Swagger comments so that they include both a summary and a description.
- Make the service configurable.
- Integrate the built-in documentation.
- Add some endpoints for defining data in the database. Feel free to let us know if you'd like some more specifics.

Once all that is done, we can work on creating the Dockerfile and Kubernetes resource definitions.

I'd like to leave this PR open at least while I'm finishing the review just to make it easier for me to keep track of what I've looked over so far. Feel free to let me know if you have any questions.

slr71 commented on internal/controllers/plans.go:
```diff
@@ -0,0 +1,39 @@
+//   200: plansResponse
+//   500: RootResponse
+
+func (s Server) GetPLansForID(ctx echo.Context) error {
```
Looks like there's a typo here too:

```suggestion
func (s Server) GetPlansForID(ctx echo.Context) error {
```

slr71 commented on internal/controllers/root.go:
```diff
@@ -0,0 +1,30 @@
+	"gorm.io/gorm"
+)
+
+// API defines the REST API of the notifications service
```
This comment looks like it was unmodified after being copied from the notifications service.

```suggestion
// Server defines the REST API of the QMS
```

slr71 commented on internal/controllers/plans.go:
```diff
@@ -0,0 +1,39 @@
+	if err != nil {
+		return ctx.JSON(http.StatusInternalServerError, model.ErrorResponse(err.Error(), http.StatusInternalServerError))
+	}
+	return ctx.JSON(http.StatusOK, model.SuccessResponse(data, http.StatusOK))
```
This produces a bit of a confusing result when I try to get information about a plan that doesn't exist:

```
$ curl -s http://localhost:9000/v1/plans/0acc725a-16bf-49c0-8831-12f055094c18 | jq .
{
  "result": {
    "ID": "0acc725a-16bf-49c0-8831-12f055094c18",
    "Name": "",
    "Description": ""
  },
  "status": "OK"
}
```

I would expect the endpoint to return a `404` status with a message indicating that the plan wasn't found. (It currently returns a `200` status along with the message above.) I checked the database, and the new plan wasn't inserted.

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,171 @@
+	data := []model.Users{}
+	err := s.GORMDB.Debug().Find(&data).Error
+	if err != nil {
+		return ctx.JSON(http.StatusBadRequest, map[string]interface{}{"error": err.Error()})
```
I'd like to see `model.ErrorResponse` used for this instead. Using a map makes it more likely that typos in field names might go undetected.

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,171 @@
+func (s Server) GetUserPlanDetails(ctx echo.Context) error {
+	username := ctx.Param("username")
+	if username == "" {
+		return ctx.JSON(http.StatusBadRequest, model.ErrorResponse("Invalid Usernaem", http.StatusBadRequest))
```
Looks like there's a typo here as well:

```suggestion
		return ctx.JSON(http.StatusBadRequest, model.ErrorResponse("Invalid Username", http.StatusBadRequest))
```

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,171 @@
+	return ctx.JSON(http.StatusOK, model.SuccessResponse(data, http.StatusOK))
+}
+
+type PlanDetails struct {
```
Public types should have comments that begin with the type name as well. `golint` will complain about this too.

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,171 @@
+	where
+	user_plans.effective_start_date<=? and
+	user_plans.effective_end_date>=? and
+	users.username=?`, now+"T00:00:00", now+"T23:59:59", username).Scan(&plandata).Error
```
It's probably easier to take care of day boundaries in SQL. You can do that using the CAST function:

```
user_plans.effective_start_date <= cast(now() as date) and
user_plans.effective_end_date >= cast(now() as date)
```

You can also simplify it a bit by using the `BETWEEN` predicate:

```
cast(now() as date) between user_plans.effective_start_date and user_plans.effective_end_date
```

You don't need to worry about adding a day to the timestamp for the effective end date comparison.

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,171 @@
+		return ctx.JSON(http.StatusInternalServerError, model.ErrorResponse(err.Error(), http.StatusInternalServerError))
+	}
+	return ctx.JSON(http.StatusOK, model.SuccessResponse(plandata, http.StatusOK))
+}
```
Instead of returning a flat list of objects, I'd really like this endpoint to return a nested structure. For example, something like this might work:

```json
{
    "plan": {
        "id": "3ee324af-aff0-4c1a-b1de-bc237824cd66",
        "name": "premier",
        "description": "The plan to have for CyVerse power users."
    },
    "effective_start_date": "2022-01-01",
    "effective_end_date": "2022-12-31",
    "usage": [
        {
            "resource_type": "Storage",
            "quota": {
                "value": 100,
                "unit": "Terabytes",
            },
            "usage": {
                "value": 34.5,
                "unit": "Terabytes",
            }
        },
        {
            "resource_type": "CPU",
            "quota": {
                "value": 1000,
                "unit": "CPU Hours",
            },
            "usage": {
                "value": 100,
                "unit": "CPU Hours",
            }
        }
    ]
}
```

This is just an initial suggestion; I wouldn't suggest modifying this endpoint just yet. We should probably have a meeting to discuss exactly what details should be returned by each endpoint in the API.

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,171 @@
+	where
+	users.username=?`+resourceFilter, username).Scan(&plandata).Error
+	if err != nil {
+		return ctx.JSON(http.StatusBadRequest, map[string]interface{}{"error": err.Error()})
```
I'd like to see `model.ErrorResponse` used here instead as well.

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,171 @@
+func (s Server) GetUserUsageDetails(ctx echo.Context) error {
+	resource := ctx.QueryParam("resource")
+	resourceFilter := ""
+	if resource != "" {
+		resourceFilter = `and resource_types.name=?`
+	}
```
Something like this is where things like [Masterminds/squirrel](https://github.com/Masterminds/squirrel) can really come in handy. I haven't used GORM, much, but I wonder if GORM has a nice way to conditionally add filters like this.

slr71 commented on internal/controllers/resourcesController.go:
```diff
@@ -0,0 +1,17 @@
+package controllers
```
This file could probably be renamed to `resources.go`.

slr71 commented on internal/db/db.go:
```diff
@@ -0,0 +1,54 @@
+
+	wrapMsg := "unable to initialize the database"
+
+	/*conn, err := sql.Open(driverName, databaseURI)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, wrapMsg)
+	}
+
+	if err := conn.Ping(); err != nil {
+		return nil, nil, errors.Wrap(err, wrapMsg)
+	}
+	conn.SetMaxIdleConns(5)
+	conn.SetMaxOpenConns(30)
+	conn.SetConnMaxLifetime(time.Hour)
+	*/
+
```
I'd like to remove all commented-out code before merging the PR.

slr71 commented on internal/db/db.go:
```diff
@@ -0,0 +1,54 @@
+		return nil, nil, errors.Wrap(err, wrapMsg)
+	}
+
+	err = MigrateTables(gormdb)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, wrapMsg)
+	}
+	return conn, gormdb, nil
```
This feature of GORM is really nice! 👍

slr71 commented on internal/db/gorm.go:
```diff
@@ -0,0 +1,66 @@
+	gormDB, err := gorm.Open(postgres.New(postgres.Config{
+		Conn: db,
+	}), &gorm.Config{
+		DisableForeignKeyConstraintWhenMigrating: true,
```
No foreign keys are being added to the database at all. That makes me wary.

gdprasadAU replied:
We I am wondering why it is not being added.
I will try to figure it out.

slr71 commented on internal/model/plan.go:
```diff
@@ -0,0 +1,93 @@
+	Name string `gorm:"column:name;type:varchar(100)"`
+	// the Description for the plan
+	//required: true
+	Description string `gorm:"column:description;type:text"`
```
A Plan has many PlanQuotaDefaults, so PlanQuotaDefaults should be referenced here as well. See [this documentation](https://gorm.io/docs/has_many.html) for details.

I don't think that anyone on the team is very familiar with GORM, so what we might want to do is to have a few of us get together to work through a few examples. It might be nice to be able to play with this to see if we can get it to do what we want.

gdprasadAU committed:
make file changes

gdprasadAU committed:
Makefile changes

gdprasadAU committed:
Merge branch 'GDTest' of https://github.com/cyverse/QMS into GDTest

gdprasadAU committed:
Makefile changes

gdprasadAU committed:
Typos and working on comments.

gdprasadAU committed:
Added Gorm foreign keys for plans

slr71 committed:
CORE-1670: restructure the code a little to change the executable name

slr71 committed:
Merge pull request #3 from cyverse/deploy

CORE-1670: restructure the code a little to change the executable name

slr71 committed:
CORE-1670: updated dependencies

gdprasadAU committed:
Made foreign changes and added endpoints to add data.

gdprasadAU committed:
Made foreign changes and added endpoints to add data.

slr71 committed:
CORE-1670: cleaned up some files that didn't need to be included in source control

slr71 committed:
CORE-1670: more Makefile changes that I missed earlier

slr71 committed:
CORE-1670: added the Dockerfile

slr71 committed:
CORE-1670: moved the log namespace to github.com/cyverse-de/echo-middleware

gdprasadAU committed:
Added Route from admin to update plan of user with username

slr71 committed:
CORE-1670: made the database URI configurable

slr71 committed:
CORE-1670: turn off the report-caller feature in logrus until the log middleware can be updated to support it

slr71 committed:
CORE-1670: fixed a database initialization error and renamed the Plan struct

slr71 committed:
CORE-1670: added default values for all ID columns

gdprasadAU committed:
Added Route to add Quotas

gdprasadAU committed:
Merge pull request #4 from cyverse/deploy

CORE-1670: prepare for deploying the QMS service

slr71 committed:
CORE-1670: initialized the repository for skaffold

slr71 committed:
CORE-1670: added some GitHub actions

CORE-1670: fixed some GitHub action warnings

CORE-1670: removed a job step that I forgot to remove earlier

CORE-1670: updated the GitHub actions to remove the deploy branch

slr71 committed:
CORE-1670: removed the skaffold-build GitHub action for now

gdprasadAU committed:
Added to add data into the database.

slr71 commented on internal/controllers/admin.go:
```diff
@@ -0,0 +1,231 @@
+	if err != nil {
+		if strings.Contains(err.Error(), "invalid input") {
+			return ctx.JSON(http.StatusBadRequest, model.ErrorResponse(err.Error(), http.StatusBadRequest))
+		}
```
Having invalid input in the database might not necessarily mean that there was invalid input from the user, so I think it'll be safe to leave this as a 500 error anytime we get an error response from the database. If possible, the code should probably check for any input that could cause a database error in advance.

slr71 commented on internal/controllers/admin.go:
```diff
@@ -0,0 +1,231 @@
+	}
+
+	value := req.Value
+	if req.Type == "sub" {
```
The `"sub"` string should probably be in a string constant somewhere to avoid the risk of tpyos.

slr71 commented on internal/controllers/admin.go:
```diff
@@ -0,0 +1,231 @@
+
+func (s Server) UpdateUsages(ctx echo.Context) error {
+	req := UpdateUsagesReq{}
+	fmt.Println(req, "================================>")
```
If these are debugging statements that you don't intend to retain then these should be removed when you don't need them anymore. Otherwise, they should be converted to log statements so that the formatting will match the rest of the log entries.

slr71 commented on internal/controllers/admin.go:
```diff
@@ -0,0 +1,231 @@
+	fmt.Println(req, "================================>")
+
+	err := ctx.Bind(&req)
+	fmt.Println(req, "================================>")
```
This also appears to be debugging code that can be removed.

slr71 commented on internal/controllers/admin.go:
```diff
@@ -0,0 +1,231 @@
+	resourcefilter := ""
+	if resource != "" {
+		resourcefilter = ` and resource_types.name = '` + resource + `'`
+	}
```
Using string concatenation to build SQL is dangerous. This should probably be refactored to use a parameter instead. GORM also has its own SQL builder, so I'd recommend using that.

slr71 commented on internal/controllers/admin.go:
```diff
@@ -0,0 +1,231 @@
+	usernamefilter := ""
+	if username != "" {
+		usernamefilter = ` and users.username = '` + username + `'`
+	}
```
Same here; using an SQL builder or positional parameters would be better.

slr71 commented on internal/controllers/admin.go:
```diff
@@ -0,0 +1,231 @@
+	join users on users.id = user_plans.user_id
+	where
+	user_plans.effective_start_date <=? and
+	user_plans.effective_end_date >=? `+usernamefilter+resourcefilter, now+"T00:00:00", now+"T23:59:59").Scan(&plandata).Error
```
You can just compare the effective start and end dates to the current date and time. Postgres has a CURRENT_TIMESTAMP variable that you can use. Adjusting the comparisons for day boundaries isn't necessary for something like this.

slr71 commented on internal/controllers/admin.go:
```diff
@@ -0,0 +1,231 @@
+	usernamefilter := ""
+	if username != "" {
+		usernamefilter = ` and users.username = '` + username + `'`
+	}
```
Once again, avoid using string concatenation to build SQL statements.

slr71 commented on internal/controllers/admin.go:
```diff
@@ -0,0 +1,231 @@
+		join resource_types on resource_types.id=quotas.resource_type_id
+		join users on users.id = user_plans.user_id
+		where 
+		cast(now() as date) between user_plans.effective_start_date and user_plans.effective_end_date` + usernamefilter + resourcefilter).Scan(&plandata).Error
```
This is a good way to do the date comparison. 👍

slr71 commented on internal/controllers/admin.go:
```diff
@@ -0,0 +1,231 @@
+	where
+	user_plans.effective_start_date<=? and
+	user_plans.effective_end_date>=? and
+	users.username=?`, now+"T00:00:00", now+"T23:59:59", username).Scan(&plandata).Error
```
This is another spot where the date comparison can be simplified.

slr71 commented on internal/controllers/admin.go:
```diff
@@ -0,0 +1,231 @@
+// 		return ctx.JSON(http.StatusInternalServerError, model.ErrorResponse(err.Error(), http.StatusInternalServerError))
+// 	}
+// 	return ctx.JSON(http.StatusOK, model.SuccessResponse("Success", http.StatusOK))
+// }
```
We'd like to have all commented-out code removed before merging the request if possible.

slr71 commented on internal/controllers/plans.go:
```diff
@@ -0,0 +1,155 @@
+}
+
+func (s Server) AddPlans(ctx echo.Context) error {
+
```
The blank line here should probably be removed for consistency.

slr71 commented on internal/controllers/plans.go:
```diff
@@ -0,0 +1,155 @@
+	Name string `json:"name"`
+}
+
+func (s Server) AddPlans(ctx echo.Context) error {
```
This appears to be adding just a single plan. If so, the function name should be `addPlan`.

slr71 commented on internal/controllers/plans.go:
```diff
@@ -0,0 +1,155 @@
+	}
+	return ctx.JSON(http.StatusOK, model.SuccessResponse("Success", http.StatusOK))
+}
+func (s Server) AddResourceType(ctx echo.Context) error {
```
There should be a blank line before this function definition.

slr71 commented on internal/controllers/plans.go:
```diff
@@ -0,0 +1,155 @@
+}
+func (s Server) AddResourceType(ctx echo.Context) error {
+	// id := "230d8bd2-7cc5-11ec-90d6-0242ac120003"
+
```
These lines should be removed.

slr71 commented on internal/controllers/plans.go:
```diff
@@ -0,0 +1,155 @@
+func (s Server) AddPlanQuotaDefault(ctx echo.Context) error {
+	id := "230d8bd2-7cc5-11ec-90d6-0242ac120003"
+	planID := "2e146110-7bf1-11ec-90d6-0242ac120003"
+	resourceTypeID := "1783e71c-7cb5-11ec-90d6-0242ac120003"
```
It seems odd to have IDs hard-coded in here. Is this debugging code?

slr71 commented on internal/controllers/plans.go:
```diff
@@ -0,0 +1,155 @@
+	}
+	return ctx.JSON(http.StatusOK, model.SuccessResponse("Success", http.StatusOK))
+}
+func (s Server) AddPlanQuotaDefault(ctx echo.Context) error {
```
Another case where a function definition should be preceded by a blank line.

slr71 commented on internal/controllers/plans.go:
```diff
@@ -0,0 +1,155 @@
+}
+
+func (s Server) AddUserPlanDetails(ctx echo.Context) error {
+
```
Another blank line that could probably be removed.

slr71 commented on internal/controllers/plans.go:
```diff
@@ -0,0 +1,155 @@
+	id := "230d8bd2-7cc5-11ec-90d6-0242ac120003"
+	planID := "2e146110-7bf1-11ec-90d6-0242ac120003"
+	resourceTypeID := "1783e71c-7cb5-11ec-90d6-0242ac120003"
+	var req = model.PlanQuotaDefaults{ID: &id, PlanID: &planID, ResourceTypeID: &resourceTypeID, QuotaValue: 1000}
```
The quota value is hard-coded here. Is this more debugging/testing code?

slr71 commented on internal/controllers/plans.go:
```diff
@@ -0,0 +1,155 @@
+}
+
+func (s Server) AddQuota(ctx echo.Context) error {
+	// id := "6b858690-7cd8-11ec-90d6-0242ac120003"
```
More commented-out code to be removed.

slr71 commented on internal/controllers/plans.go:
```diff
@@ -0,0 +1,155 @@
+	s.GORMDB.Debug().Find(&userPlan, "user_id=?", userID)
+	userPlanId := *userPlan.ID
+
+	var req = model.Quotas{UserPlanID: &userPlanId, Quota: 1000, ResourceTypeID: &resourceID}
```
There's another hard-coded quota here.

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,229 @@
+	if username == "" {
+		return ctx.JSON(http.StatusBadRequest, model.ErrorResponse("Invalid UserName", http.StatusBadRequest))
+	}
+	// now := time.Now().Format("2006-01-02")
```
More commented-out code to be removed.

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,229 @@
+	resourceFilter := ""
+	if resource != "" {
+		resourceFilter = `and resource_types.name=?`
+	}
```
This isn't risky from a security point of view, but it would be nice to use an SQL builder to avoid explicit string concatenation in here altogether.

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,229 @@
+	return ctx.JSON(http.StatusOK, model.SuccessResponse(plandata, http.StatusOK))
+}
+
+func (s Server) AddUsers(ctx echo.Context) error {
```
This should probably be renamed to `addUser`

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,229 @@
+	return ctx.JSON(http.StatusOK, model.SuccessResponse("Success", http.StatusOK))
+}
+
+func (s Server) UpdateUserQuota(ctx echo.Context) error {
```
This function appears to be updating the user's plan rather than one of the user's quota values.

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,229 @@
+	}
+	return ctx.JSON(http.StatusOK, model.SuccessResponse("Success", http.StatusOK))
+}
+func (s Server) AddUsages(ctx echo.Context) error {
```
This is another case where the function name should probably be renamed. There's also no whitespace above the function definition.

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,229 @@
+	return ctx.JSON(http.StatusOK, model.SuccessResponse("Success", http.StatusOK))
+}
+func (s Server) AddUsages(ctx echo.Context) error {
+	userPlanID := "230d8bd2-7cc5-11ec-90d6-0242ac120003"
```
Another hard-coded ID. Is this debugging code?

gdprasadAU replied:
Yes, it is debugging code.
Changed it.

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,229 @@
+}
+func (s Server) AddUsages(ctx echo.Context) error {
+	userPlanID := "230d8bd2-7cc5-11ec-90d6-0242ac120003"
+	var req = model.Usages{Usage: 5000, AddedBy: "Admin", UserPlanID: userPlanID, ResourceTypeID: "230d8bd2-7cc5-11ec-90d6-0242ac120003"}
```
More hard-coded values are in this line.

slr71 commented on internal/model/plan.go:
```diff
@@ -0,0 +1,79 @@
+	//The id for the UserPlans.
+	// in: path
+	//required: true
+	ID *string `gorm:"type:uuid;default:uuid_generate_v1()" json:"id"`
+	//The effective start date of the plan for the user.
+	// in: path
+	//required: false
+	EffectiveStartDate time.Time `gorm:"json:effective_start_date;type:date"`
+	//The effective end date of the plan for the user.
+	// in: path
+	//required: false
+	EffectiveEndDate time.Time `gorm:"json:effective_end_date;type:date"`
+	//Added by:.
+	// in: path
+	//required: false
+	AddedBy string `gorm:"json:added_by;type:varchar(100)"`
+	//The last modified of the plan for the user.
+	// in: path
+	//required: false
+	LastModifiedBy string `gorm:"json:last_modified_by;type:varchar(100)"`
```
Having `json` inside the type annotations for GORM seems odd. This seems to be an error, but we're not entirely sure. Is this a feature of GORM or were these intended to be JSON field names?

gdprasadAU replied:
Yes.

slr71 commented on internal/model/plan.go:
```diff
@@ -0,0 +1,79 @@
+
+// PlanQuotaDefaults define the structure for an Api Plan and Quota.
+// swagger:model
+type PlanQuotaDefaults struct {
```
If you rename this model to `PlanQuotaDefault` then GORM will automatically name the table `plan_quota_defaults`. That will also allow you to remove the `TableName` method.

slr71 commented on internal/model/plan.go:
```diff
@@ -0,0 +1,79 @@
+
+// UserPlans define the structure for the API User plans.
+// swagger:model
+type UserPlans struct {
```
This is another model name that is plural, but should be singular.

slr71 commented on internal/model/quotas.go:
```diff
@@ -0,0 +1,41 @@
+
+// Quotas define the structure for an Api Plan and Quota.
+// swagger:model
+type Quotas struct {
```
This model name can be singular as well.

slr71 commented on internal/model/quotas.go:
```diff
@@ -0,0 +1,41 @@
+	//Added by.
+	// in: path
+	//required: true
+	AddedBy string `gorm:"json:added_by"`
+	//The last date the record was modified
+	// in: path
+	//required: false
+	LastModifiedBy string `gorm:"json:last_modified_by"`
+	//The userPlanId for the Quota.
+	// in: path
+	//required: true
```
These two type annotations seem a little off as well.

slr71 commented on internal/model/resources.go:
```diff
@@ -0,0 +1,22 @@
+
+// ResourceTypes define the structure for ResourceTypes.
+// swagger:model
+type ResourceTypes struct {
```
This is another model name that should probably be singular.

slr71 commented on internal/model/updates.go:
```diff
@@ -0,0 +1,43 @@
+	"gorm.io/gorm"
+)
+
+type UpdateOperations struct {
```
This should probably also be singular.

slr71 commented on internal/model/updates.go:
```diff
@@ -0,0 +1,43 @@
+	return "update_operations"
+}
+
+type TrackedMetrics struct {
```
This should probably be singular as well.

slr71 commented on internal/model/updates.go:
```diff
@@ -0,0 +1,43 @@
+	return "tracked_metrics"
+}
+
+type Updates struct {
```
This should probably be singular as well..

slr71 commented on internal/model/updates.go:
```diff
@@ -0,0 +1,43 @@
+	ValueType        string    `gorm:"not null;unique" json:"value_type"`
+	Value            float64   `gorm:"not null;unique" json:"value"`
+	UpdatedBy        string    `gorm:"not null" json:"updated_by"`
+	EffectiveDate    time.Time `gorm:"json:effective_date;type:date"`
+	LastModifiedBy   string    `gorm:"json:last_modified_by;varchar(100)"`
```
These type annotations seem a little off as well.

slr71 commented on internal/model/usages.go:
```diff
@@ -0,0 +1,41 @@
+
+// Usages define the structure for API Usages.
+// swagger:model
+type Usages struct {
```
This model name should probably be singular as well.

slr71 commented on internal/model/users.go:
```diff
@@ -0,0 +1,18 @@
+
+// Users define the structure of an User.
+// swagger:model
+type Users struct {
```
This should probably be singular as well.

slr71 commented on internal/swaggerDocs/SwaggerDocs.go:
```diff
@@ -0,0 +1,77 @@
+// Package api QMS
```
We normally have `snake_case` file names in Go, so this file should be renamed to `swagger_docs.go`.

slr71 commented on server/router.go:
```diff
@@ -0,0 +1,103 @@
+	// Add middleware.
+	e.Use(middleware.Logger())
+	e.Use(middleware.Recover())
+	// e.Use(redoc.Serve(redoc.Opts{Title: "DE Administrative Requests API Documentation"}))
+	// e.Logger.Info("loading service information")
+	// serviceInfo, err := getSwaggerServiceInfo()
+	// if err != nil {
+	// 	e.Logger.Fatal(err, serviceInfo)
+	// }
```
Also, any commented out code should be removed before we merge this.

slr71 commented on server/router.go:
```diff
@@ -0,0 +1,103 @@
+	v1.POST("/:user_name/Userplan/:plan_name/add", s.AddUserPlanDetails)
+	v1.POST("users/:user_name/quotas/add", s.AddQuota)
+	v1.POST("/usages/add", s.AddQuota)
+
```
There's an extraneous blank line here.

slr71 commented on server/router.go:
```diff
@@ -0,0 +1,103 @@
+
+	// view all the plan details along with usage and quotas
+	users.GET("/:username/quotas", s.GetUserAllQuotas)
+	// users.GET("/:username/quotas/:quotaId", s.GetUserQuotaDetails)
+	// what if the users have multile palms and wants to look at a particular plan
+	// users.GET("/:username/usages", s.GetUserUsageDetails)
```
In general, it's a good idea to remove commented-out code.

slr71 commented on server/router.go:
```diff
@@ -0,0 +1,103 @@
+	//Users
+	//
+	users := v1.Group("/users")
+	users.POST("/:user_name/add", s.AddUsers)
+
+	users.GET("/:username/plan", s.GetUserPlanDetails)
+
```
The variable placeholder names are a bit inconsistent here. These should probably all use `snake_case`.

slr71 commented on server/router.go:
```diff
@@ -0,0 +1,103 @@
+	v1.GET("/resources", s.GetAllResources)
+	v1.POST("/resources/:resource_name/:resource_unit/add", s.AddResourceType)
+	v1.POST("/planquota/add", s.AddPlanQuotaDefault)
+	v1.POST("/:user_name/Userplan/:plan_name/add", s.AddUserPlanDetails)
```
Could you explain the intent of this endpoint? Just from the name, it sounds like it's adding details to the user plan, but what details is it adding?

gdprasadAU replied:
Updating a particular user's plan in the userplan table.

slr71 commented on server/router.go:
```diff
@@ -0,0 +1,103 @@
+	//Rersources
+	v1.GET("/resources", s.GetAllResources)
+	v1.POST("/resources/:resource_name/:resource_unit/add", s.AddResourceType)
+	v1.POST("/planquota/add", s.AddPlanQuotaDefault)
```
This endpoint appears to add a default quota to a plan, but there's nothing in the URL identifying which plan should be updated. For something like this, it's probably best to have something in the URL that identifies the entity that is being updated.

slr71 reviewed (commented):
We made some more suggestions for the code. One thing that was difficult to add to the in-line comments was the presence of a `.DS_Store` file checked into source control. This file should be removed. It might be a good idea to add `.DS_Store` to `.gitignore` so that these files aren't checked in in the future.

slr71 reviewed (changes requested):
There are some things in here that we'd like to see fixed before the changes are merged into the main branch.

slr71 commented:
Oh, one thing that I meant to mention earlier was that the `swaggerDocs` package should be renamed because Go coding standards recommend against having mixed-case package names. It could be renamed `swaggerdocs` or simply `swagger`. I tend to prefer the latter.

gdprasadAU committed:
Merge pull request #5 from cyverse/deploy

CORE-1670: added the deployment files

Commit:
Format query and code in GetAllActiveUsage(), change reference to users.username to users.user_name

gdprasadAU committed:
Testing and working on code.

gdprasadAU committed:
Merge branch 'GDTest' of https://github.com/cyverse/QMS into GDTest

Commit:
Result of running 'go mod tidy'

Commit:
Fix updating usages

Commit:
Fix typo in the GetAllActiveUsage() query

Commit:
Update UpdateUsages()

Commit:
Remove debug print statement

gdprasadAU committed:
Merge pull request #6 from cyverse/johnw

Updates to usages controllers.

gdprasadAU committed:
Merge

gdprasadAU committed:
Made required changes as requested.

gdprasadAU committed:
Removed Commented code.

slr71 commented on Dockerfile:
```diff
@@ -0,0 +1,20 @@
+FROM golang:1.17
+
+RUN go install github.com/jstemmer/go-junit-report@latest
```
TO-DO: remove go-junit.

slr71 commented on internal/controllers/plans.go:
```diff
@@ -0,0 +1,178 @@
+}
+
+func (s Server) AddPlanQuotaDefault(ctx echo.Context) error {
+	planName := "Basic"
+	resourceName1 := "CPU"
+	cpuValue := 4.00
+	resourceName2 := "STORAGE"
+	storageValue := 1000.00
```
These hard-coded values should be replaced with either constants or information from the incoming request.

slr71 commented on internal/controllers/plans.go:
```diff
@@ -0,0 +1,178 @@
+	resourceName2 := "STORAGE"
+	storageValue := 1000.00
+	var plan = model.Plan{Name: planName}
+	s.GORMDB.Debug().Find(&plan, "name=?", planName)
```
This isn't a blocker, but `s.GORMDB` and `s.GORMDB.Debug()` are the same type, so it would be possible to store the result of calling `s.GORMDB.Debug()` in the server struct in order to enable debugging everywhere. This would  be a bit cleaner while still allowing us to easily disable the debugging.

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,171 @@
+	if err != nil {
+		return ctx.JSON(http.StatusInternalServerError, model.ErrorResponse(err.Error(), http.StatusInternalServerError))
+	}
+	storage := "STORAGE"
+	cpu := "CPU"
```
This isn't a blocker, but these should probably be constants.

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,171 @@
+	s.GORMDB.Debug().Find(&cpuResource, "name=?", cpu)
+	cpuId := *cpuResource.ID
+	userPlanId := *userPlan.ID
+	var defaultStorageQuota = model.PlanQuotaDefault{}
```
This isn't a blocker, but it's possible to combine all of the separate `var` statements in a function into a single `var` block at the beginning of the function.

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,171 @@
+	var defaultcpuQuota = model.PlanQuotaDefault{}
+	s.GORMDB.Debug().Find(&defaultcpuQuota, "resource_type_id=?", cpuId)
+	defaultCPUQuotaValue := defaultcpuQuota.QuotaValue
+	var userQuota = []model.Quota{{AddedBy: "Admin", UserPlanID: &userPlanId, ResourceTypeID: &storageId, Quota: defaultStorageQuotaValue}, {AddedBy: "Admin", UserPlanID: &userPlanId, ResourceTypeID: &cpuId, Quota: defaultCPUQuotaValue}}
```
This isn't a blocker either, but this line should probably be formatted so that the data structure spans multiple lines.

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,171 @@
+}
+
+func (s Server) UpdateUserPlan(ctx echo.Context) error {
+	planname := ctx.Param("plan_name")
+	if planname == "" {
+		return ctx.JSON(http.StatusBadRequest, model.ErrorResponse("Invalid Plan Name", http.StatusBadRequest))
+	}
+	username := ctx.Param("user_name")
+	if username == "" {
+		return ctx.JSON(http.StatusBadRequest, model.ErrorResponse("Invalid UserName", http.StatusBadRequest))
+	}
+	var user = model.User{UserName: username}
+	s.GORMDB.Debug().Find(&user, "user_name=?", username)
+	userID := *user.ID
+	var plan = model.Plan{Name: planname}
+	s.GORMDB.Debug().Find(&plan, "name=?", planname)
+	planId := *plan.ID
+	var userPlan = model.UserPlan{}
+	s.GORMDB.Debug().Find(&userPlan, "user_id=?", userID)
+	userPlanId := *userPlan.ID
+	if *userPlan.PlanID == planId {
+		return ctx.JSON(http.StatusInternalServerError, model.ErrorResponse("User cannot be updated with the existing plan: "+planname, http.StatusInternalServerError))
+	}
+	errdb := s.GORMDB.Debug().Model(&userPlan).Where("id=?", userPlanId).Update("plan_id", planId).Error
+	if errdb != nil {
+		return ctx.JSON(http.StatusInternalServerError, model.ErrorResponse(errdb.Error(), http.StatusInternalServerError))
+	}
+	return ctx.JSON(http.StatusOK, model.SuccessResponse("Success", http.StatusOK))
```
This also isn't a blocker, but I like to have blank lines interspersed in function definitions so that related lines are grouped together. That makes it easier to reason about the function.

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,171 @@
+		return ctx.JSON(http.StatusBadRequest, model.ErrorResponse("Invalid Resource Name", http.StatusBadRequest))
+	}
+	var res = Result{}
+	err := s.GORMDB.Table("user_plans").Select("user_plans.id, users.user_name,plan_quota_defaults.resource_type_id ").Joins("JOIN users on user_plans.user_id=users.id").Joins("JOIN plan_quota_defaults on plan_quota_defaults.plan_id=user_plans.plan_id").Joins("JOIN resource_types on resource_types.id=plan_quota_defaults.resource_type_id").Where("users.user_name=? and resource_types.name=?", username ...
```
```suggestion
	err := s.GORMDB.Table("user_plans").
		Select("user_plans.id, users.user_name,plan_quota_defaults.resource_type_id ").
		Joins("JOIN users on user_plans.user_id=users.id").
		Joins("JOIN plan_quota_defaults on plan_quota_defaults.plan_id=user_plans.plan_id").
		Joins("JOIN resource_types on resource_types.id=plan_quota_defaults.resource_type_id").
		Where("users.user_name=?", username).
		Where("resource_types.name=?", resourceName).
		Scan(&res).
		Error
```

slr71 commented on internal/controllers/users.go:
```diff
@@ -0,0 +1,171 @@
+	if err != nil {
+		return ctx.JSON(http.StatusInternalServerError, model.ErrorResponse(err.Error(), http.StatusInternalServerError))
+	}
+	var req = model.Usage{Usage: 5000, AddedBy: "Admin", UserPlanID: res.ID, ResourceTypeID: res.ResourceTypeId}
```
Also not a blocker, but hard-coded values should be removed.

slr71 reviewed (approved):
Looks good! There were still a few things to mention, but nothing that can't be fixed later. I think we're ready to get this deployed into our QA environment.
